subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy

```diff
@@ -223,7 +217,82 @@ class MissingTaskDependenciesIntegrationTest extends AbstractIntegrationSpec {
         """
 
         expect:
-        executer.expectDocumentedDeprecationWarning(DEPRECATION_WARNING)
+        expectMissingDependencyDeprecation(":producer", ":consumer")
         succeeds("producer", "consumer")
     }
+
+    def "can have tasks which consume a filtered project directory"() {
+        file("src/main/java/MyClass.java").createFile()
+        buildFile << """
+            task firstTask {
+                def outputFile = file("build/output1.txt")
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.text = "first"
+                }
+            }
+            task lastTask {
+                def outputFile = file("build/output2.txt")
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.text = "last"
+                }
+            }
+            task packageProjectDir(type: Zip) {
+                from(project.projectDir) {
+                    include 'src/**'
+                }
+                destinationDirectory = file("build")
+                archiveBaseName = "output3"
+            }
+        """
+
+        when:
+        run("firstTask", "packageProjectDir", "lastTask")
+        then:
+        executedAndNotSkipped(":firstTask", ":packageProjectDir", ":lastTask")
+    }
+
+    def "detects filtered trees with missing dependencies"() {
+        file("src/main/java/MyClass.java").createFile()
+        buildFile << """
+            task firstTask {
+                def outputFile = file("build/output1.txt")
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.text = "first"
+                }
+            }
+            task lastTask {
+                def outputFile = file("build/output2.txt")
+                outputs.file(outputFile)
+                doLast {
+                    outputFile.text = "last"
+                }
+            }
+            task packageProjectDir(type: Zip) {
+                from(project.projectDir) {
+                    include 'build/**'
+                }
+                destinationDirectory = file("build")
+                archiveBaseName = "output3"
+            }
+        """
+
+        when:
+        expectMissingDependencyDeprecation(":firstTask", ":packageProjectDir")
+        expectMissingDependencyDeprecation(":lastTask", ":packageProjectDir")
+        run("firstTask", "packageProjectDir", "lastTask")
+        then:
+        executedAndNotSkipped(":firstTask", ":packageProjectDir", ":lastTask")
+    }
+
+    void expectMissingDependencyDeprecation(String producer, String consumer) {
+        executer.expectDocumentedDeprecationWarning(
+            "Task '${consumer}' uses the output of task '${producer}', without declaring an explicit dependency (using dependsOn or mustRunAfter) or an implicit dependency (declaring task '${producer}' as an input). " +
```

Comment:
```suggestion
            "Task '${consumer}' uses the output of task '${producer}' without declaring an explicit dependency (using Task.dependsOn() or Task.mustRunAfter()) or an implicit dependency (declaring task '${producer}' as an input). " +
```

Reply:
It is a bit weird for me that we mention `producer` twice, but not `consumer`; seems this should be symmetric.

Reply:
Another option would be to break this up into more sentences:

```suggestion
            "Task '${consumer}' consumes the output of task '${producer}' without a dependency between the two. Either consume the output of task '${producer}', or declare an explicit dependency using Task.dependsOn() or Task.mustRunAfter(). " +
```

Sorry, the message isn't perfect, but I think it conveys the direction I think we could go.

Reply:
I used your first suggestion. I think it is fine to mention `producer` twice, given that you declare the dependency on the `producer`.